torch/csrc/jit/interpreter.cpp

```diff
@@ -1,42 +1,208 @@
+        unsafeTensorBorrow(input),
+        flags,
+        handle->forward_inputs->num_inputs++,
+        handle->forward_inputs);
+    } else {
+      autograd::VarFlags flags = {false, false};
+      return autograd::make_variable(unsafeTensorBorrow(input), flags);
+    }
+  }
+  at::RefCounted* addOutput(const autograd::Variable & output) {
+    if(handle) {
+      handle->forward_outputs.emplace_back(output.grad_fn(),output.output_nr());
+    }
+    at::Tensor tensor = output.data();
+    return tensor.detach();
+  }
+  void writeTo(refcounted_list & outputs) {
+    // note: no if(handle) guard
+    // because an unused handle is still produced as an output
+    outputs.push_back(handle);
+  }
+private:
+  Handle* handle = nullptr;
+};
+
+bool hasHandleOutput(Node * n) {
+  if(n->outputs().size() == 0)
+    return false;
+  auto & last = n->outputs().back();
+  return last->isHandle() && last->uses().size() > 0; // don't bother creating a handle if it is never used
+}
+
+Operation createPythonCallback(PythonOp* op) {
+  py::object func = py::handle(op->pyobj.get()).attr("apply");
+  bool has_handle = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs, refcounted_list & outputs) {
+    AutoGIL gil;
+    py::tuple py_inputs(op->cconv.size());
+    size_t i = 0;
+    size_t next_scalar = 0;
+    size_t next_tensor = 0;
+    HandleBuilder builder(has_handle);
+    for(auto arg_type : op->cconv) {
+      if(arg_type == 's') {
+        py_inputs[i] = py::reinterpret_borrow<py::object>(op->scalar_args[next_scalar++].get());
+      } else if(arg_type == 't') {
+        py_inputs[i] = THPVariable_Wrap(
+          builder.addInput(inputs.at(next_tensor), op->var_flags.at(next_tensor)));
+        next_tensor++;
+      }
+      i++;
+    }
+    py::object py_outputs(func(*py_inputs));
+
+    auto addOutput = [&](py::handle entry) {
+      if(!THPVariable_Check(entry.ptr())) {
+        throw std::runtime_error("Function.apply returned a non-Variable output");
+      }
+      THPVariable *var = (THPVariable*) entry.ptr();
+      outputs.push_back(builder.addOutput(var->cdata));
+    };
+    if(!PyTuple_Check(py_outputs.ptr())) {
+      addOutput(py_outputs);
+    } else {
+      for(py::handle entry : py::tuple(py_outputs)) {
+        addOutput(entry);
+      }
+    }
+    builder.writeTo(outputs);
+  };
+}
+
+Operation createCppCallback(CppOp* op) {
+  std::shared_ptr<autograd::Function> func = op->fn;
+  bool has_handle = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs, refcounted_list & outputs) {
+    HandleBuilder builder(has_handle);
+    autograd::variable_list v_inputs;
+    for(size_t i = 0; i < inputs.size(); i++) {
+      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    }
+    autograd::variable_list v_outputs = func->apply(v_inputs);
+    for(auto & output : v_outputs) {
+      outputs.push_back(builder.addOutput(output));
+    }
+    builder.writeTo(outputs);
```

Comment:
It's a bit weird to use HandleBuilder even when you're not building handles, just because it can write to outputs that have nothing to do with the actual handle.

Reply:
I added a comment about this. Handles interact with how outputs/inputs are converted so it makes sense to put it one place.